s21/p5/tester.py

```diff
@@ -0,0 +1,509 @@
+    
+    # compare length
+    if abs(len(actual)-len(expected)):
+        print(f"the number of input ips: {len(expected)}, but the number of output: {len(actual)}")
+        points -= abs(len(actual)-len(expected)) * unit_points
+    
+    # compare contents
+    for i in range(len(expected)):
+        try:
+            actual_dict = actual[i]
+            expected_dict = expected[i]
+            
+            if actual_dict['ip'] != expected_dict['ip']: # ip check
+                print(f"expected ip {expected_dict['ip']}, but found {actual_dict['ip']}")
+                points -= unit_points
+            else:
+                if actual_dict['int_ip'] != expected_dict['int_ip']: # int_ip check
+                    print(f"ip: {expected_dict['ip']}, expected ip_int {expected_dict['int_ip']} but found {actual_dict['int_ip']}")
+                    points -= unit_points
+                else:
+                    if actual_dict['region'] != expected_dict['region']: # region check
+                        print(f"ip: {expected_dict['ip']}, expected region {expected_dict['region']} but found {actual_dict['region']}")
+                        points -= unit_points
+        except IndexError as e:
+            print(f"missing: {expected[i]['ip']}")
+        except Exception as e:
+            print(type(e), e)
+    
+    # optimize: faster processing on consecutive ones, criteria: < 20% of random access
+    avg_time_inconsecutive = np.mean([x['ms'] for x in actual[:6]])
+    avg_time_consecutive = np.mean([x['ms'] for x in actual[6:]])
+    # print('time', avg_time_inconsecutive, avg_time_consecutive)
+    
+    if (avg_time_inconsecutive / 10) < avg_time_consecutive:
+        print("(-10) Not enough optimized for consecutive ips! Time for consecutive ips can be less than 10% of inconsecutive ones")
+        print(f"average elapsed time for consecutive ips: {avg_time_consecutive}")
+        print(f"average elapsed time for inconsecutive ips: {avg_time_inconsecutive}")
+        points -= 10
+    return max(points, 0)
+
+@test(points=25)
+def sample():
+    points = 25
+    zname = "large.zip"
+    zout = "small_samp.zip"
+    stride = 1000
+    start_time = time.time()
+    try:
+        run("sample", zname, zout, stride) # include stride of test data set
+    except subprocess.TimeoutExpired:
+        print("Execution time for sample should be less than 90 seconds")
+        return 0
+    except Exception as e:
+        print(type(e), e)
+        
+    elapsed_time = time.time() - start_time
+    if elapsed_time > 20:
+        print("(-10) sample command can be optimized further. (<= 20secs), elapsed time:", elapsed_time)
+        print("Make sure you have an O(M+N) implementation")
+        points -= 10
+        
+    
+    # write on actual_json
+    actual_count = 0
+    for i, row in enumerate(zip_csv_iter(zout)):
+        actual_json[f"sample_{i}"] = row
+        actual_count += 1
+    
+    # dump in the middle for convenience
```

Comment:
Why do we want this if it already gets updated by is_expected?

Reply:
It looks like I was confused about something. Thanks!